### PR TITLE
`L.NegativeSampling` ChainerX support

### PR DESCRIPTION
**Pull request by niboshi**
Fixes `L.NegativeSampling` for ChainerX.

### CODE REVIEW

**Comment by niboshi** [comment]
`TestNegativeSampling.test_backward_compare_with_numpy` causes a strange failure due to invalid dtype.
I susupect that's caused from `using_config('dtype', ...)` in `setUp`/`tearDown` pair: if the test is skipped, the `tearDown` is never called.

**Comment by niboshi** [comment]
Well, now the problem is gone for some reason...

The problem I mentioned above, that `tearDown` is not called if the test is skipped, does not seem persistent. The `--pdb` option of `pytest` could possibly be the fault: without `--pdb` the teardown seems to be called no matter whether the test is skipped, but with it, the opposite.

File: tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py [hunk 1]
```diff
@@ -4,27 +4,28 @@
 
 import chainer
 from chainer import backend
-from chainer.backends import cuda
+from chainer.backends import _cpu
```
Comment (hvy) [inline tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py]: How about using `chainer.backend.CpuDevice.send` for cpu transfer preferring public APIs?

File: chainer/backends/_chainerx.py [hunk 2]
```diff
@@ -165,11 +164,7 @@ def _array_from_chainerx(array):
     if array is None:
         return None
     if not isinstance(array, chainerx.ndarray):
-        if isinstance(array, chainer.get_array_types()):
-            return array
-        raise TypeError(
-            'Tried to convert to a non-ChainerX array from an invalid type: '
-            '{}'.format(type(array)))
+        return array
```
Comment (hvy) [inline chainer/backends/_chainerx.py]: I see this behavior is convenient (for slices?) but it'll stop being symmetrical to "to_chainerx". Is it difficult to keep the old behavior and check the type before calling this function?

File: chainer/functions/array/get_item.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def check_type_forward(self, in_types):
         type_check._argname(in_types, ('x',))
 
     def forward(self, xs):
-        return utils.force_array(xs[0][self.slices]),
+        return utils.force_array(xs[0][backend.from_chainerx(self.slices)]),
```
Comment (hvy) [inline chainer/functions/array/get_item.py]: I now feel the attribute fallback https://github.com/chainer/chainer/pull/5745 wasn't robust enough. Don't you think we should extend it to support contained ndarrays too?

Comment (asi1024) [reply chainer/functions/array/get_item.py]: Beyond the scope of this PR.

**Comment by asi1024** [comment]
jenkins, test this please.

**Comment by asi1024** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
CI failure seems unrelated, LGTM (and @asi1024 ).

**Pull request merged**